### PR TITLE
Brad/exp 458 fe allow pasting in a git clone url to create project as

### PR DESCRIPTION
**Pull request by selfcontained**
## Description

The primary aim for this PR was to allow pasting in a git clone url when creating a project, and allowing the user to create w/ that clone url if there isn't a repo that matches their search, and if the url looks like a clone url. In case there are problems loading the repositories, this provides a workaround.

image

**_Notes:_**
The `NewProject.tsx`  file has a lot going on it, which made it quite difficult to modify some of the functionality w/o effecting the rest. There were like 8 useEffects() and several layers of nested rendering functions. I ended up refactoring most of that component and breaking it out into more manageable pieces, as well as prepping some of the queries/mutations to let us build an incremental repo search for this UI as a followup.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 9a8b7a2</samp>

This pull request adds several new files with TypeScript code that implement the user interface and the data fetching logic for creating projects from GitHub repositories. It defines custom React hooks for querying and mutating GitHub-related data, and React components for rendering the different elements of the new project creation flow. It also provides a utility function for reconfiguring the GitHub app if needed.

</details>

## Related Issue(s)

Fixes EXP-458

## How to test

* Try and create new projects on the Preview Environment. 
* Try pasting in git clone urls and creating that way

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - brad-exp-4a012c26733</li>
	<li><b>🔗 URL</b> - <a href="https://brad-exp-4a012c26733.preview.gitpod-dev.com/workspaces" target="_blank">brad-exp-4a012c26733.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - brad-exp-458-fe-allow-pasting-in-a-git-clone-url-to-create-project-as-gha.15508</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }, [repoSearchFilter]);
+
+    const handleCreate = useCallback(() => {
+        let name = "";
+        let slug = "";
+
+        try {
+            // try and parse the url for owner/repo path parts
+            console.log("url: ", normalizedURL.substring(0, normalizedURL.length - 4));
+            const [owner, repo] = new URL(normalizedURL.substring(0, normalizedURL.length - 4)).pathname.split("/");
```
Comment (filiptronicek) [inline components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:46]: This split produces an array like this: 
```ts
[ "", "gitpod-io", "workspace-images" ]
```

So we should filter out the first item of the array. Here's 2 different ways I can think of to do that:

```suggestion
            const [_, owner, repo] = new URL(normalizedURL.substring(0, normalizedURL.length - 4)).pathname.split("/");
```

```suggestion
            const [owner, repo] = new URL(normalizedURL.substring(0, normalizedURL.length - 4)).pathname.split("/").filter(Boolean);
```

Also, for better readability, I'd be explicit with what the substringing is for
```suggestion
            const [owner, repo] = new URL(normalizedURL.substring(0, normalizedURL.length - ".git".length)).pathname.split("/");
```

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+export const NewProjectCreateFromURL: FC<Props> = ({ repoSearchFilter, isCreating, onCreateProject }) => {
+    const { toast } = useToast();
+    const showCreateFromURL = useMemo(() => {
+        // TODO: Only accounts for https urls, need to account for ssh clone urls too?
+        const looksLikeURL = isURL(repoSearchFilter, {
+            require_protocol: true,
+            protocols: ["https"],
+        });
+
+        const hasTrailingGit = /\.git$/.test(repoSearchFilter);
```
Comment (filiptronicek) [inline components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:28]: This is also true when the URL looks like this:
```
https://example.com#.git
https://example.com?s=.git
```

It's a pretty small issue and the project creation will fail anyway, but wanted to bring it up.

**Review by filiptronicek — COMMENTED** [review]
Hey Brad, chiming in pretty early as it's still a draft, just some minor suggestions 🙏

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        return url;
+    }, [repoSearchFilter]);
+
+    const handleCreate = useCallback(() => {
+        let name = "";
+        let slug = "";
+
+        try {
+            // try and parse the url for owner/repo path parts
```
Comment (AlexTugarev) [inline components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:44]: We've a util for that in server. I could make sense to pull it out to gitpod-protocol for reuse. Have a look at the special cases, some of them actually introduced for BBS:
https://github.com/gitpod-io/gitpod/blob/de09b5f828e23b8dfb6774c750c94b2db846c00d/components/server/src/repohost/repo-url.ts

Comment (AlexTugarev) [reply components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:44]: Screenshot 2023-08-17 at 15 31 02

And this request params show, something is off.

File: components/dashboard/src/data/projects/create-project-mutation.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    const refreshProjects = useRefreshProjects();
+    const { data: org } = useCurrentOrg();
+
+    return useMutation<Project, Error, CreateProjectArgs>(
+        async ({ name, slug, cloneUrl, appInstallationId }) => {
+            if (!org) {
+                throw new Error("No org currently selected");
+            }
+
+            return await getGitpodService().server.createProject({
```
Comment (AlexTugarev) [inline components/dashboard/src/data/projects/create-project-mutation.ts:25]: (nothing wrong with that line, just don't know where to attach.)

The current check for repository permissions is case-sensitive. This is why the `createProject` will fail, as it's transformed to lower case. Let me see if I can fix the permission check real quick in the other PR. 

Screenshot 2023-08-17 at 15 41 12
Screenshot 2023-08-17 at 15 41 53

**Comment by AlexTugarev** [comment]
@selfcontained, in general this is almost perfect! 

Besides the other comments, I think the only issue I see is with discoverability of this feature. For testing, I didn't look too much into the FE code, but just tried to use it. Here are my thoughts in short:
* First there are no hints about the feature of pasting in the cloneURL
* When trying to enter a GH repo URL, I still saw the "No results" rendering 
  * This is where I would have expected to see a hint. Maybe we can think of parsing for a valid URL in general, and render a hint if it's not a complete clone URL yet, so that one could understand the expectation at this point.

I leave it up to you to decide on how far you want to take it in here, or create follow-up issues.

File: components/dashboard/src/data/projects/create-project-mutation.ts (hunk 4 above)
Comment (AlexTugarev) [reply components/dashboard/src/data/projects/create-project-mutation.ts:25]: It's resolved here: https://github.com/gitpod-io/gitpod/pull/18532/commits/1015169880c8bd170d38dc8b79422589c2185f8b

Please rebase

**Comment by filiptronicek** [comment]
@selfcontained whenever I try adding a repo from a clone URL, I seem to get an error from the GitHub API, preventing the Project success page from showing up, even though it is created in the background

image
image

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx [hunk 5]
```diff
@@ -59,14 +59,14 @@ export const NewProjectCreateFromURL: FC<Props> = ({ repoSearchFilter, isCreatin
...
 
-            if (!owner && !repo) {
+            if (!repo) {
                 throw new Error();
             }
 
-            name = repo || owner;
-            slug = (repo || owner).toLowerCase();
+            name = repo;
+            slug = [repo, owner].filter(Boolean).join("-").toLowerCase();
```
Comment (AlexTugarev) [inline components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:69]: just to resurface: this won't work nice with personal BBS repos, e.g. `https://bitbucket.gitpod-dev.com/scm/~alex/personal-repo.git`

**Review by AlexTugarev — APPROVED** [review]
It worked nicely 👍🏻 

🙇🏻‍♂️ Awesome work!

**Review by filiptronicek — APPROVED** [review]
Just re-tried; works great!

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx (hunk 5 above)
Comment (selfcontained) [reply components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx]: I ended up swapping it so it's repo at the end. So in this case the slug would be `scm-~alex-personal-repo` - is that a problem?

File: components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx (hunk 3 above)
Comment (selfcontained) [reply components/dashboard/src/projects/new-project/NewProjectCreateFromURL.tsx:55]: Thanks for catching, it's updated to use the repo name correctly now, and the slug becomes a mashup of the "owner" path parts and the repo name.

image

**Pull request merged**